prometheus-sachet: init at 0.3.1

GovanifY opened the pull request:
###### Motivation for this change

This adds [sachet](https://github.com/messagebird/sachet) to nixpkgs, an alertmanager SMS alerting backend, which can be useful for paging.

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

GovanifY committed:
prometheus-sachet: init at 0.3.1

SuperSandro2000 commented on pkgs/servers/monitoring/prometheus/sachet.nix:
```diff
@@ -0,0 +1,26 @@
+, buildGoModule
+, fetchFromGitHub
+}:
+{
+  prometheus-sachet = buildGoModule rec {
```
```suggestion

buildGoModule rec {
```

SuperSandro2000 commented on pkgs/servers/monitoring/prometheus/sachet.nix:
```diff
@@ -0,0 +1,26 @@
+
+    src = fetchFromGitHub {
+      owner = "messagebird";
+      repo = "${pname}";
+      rev = "${version}";
```
```suggestion
      repo = pname;
      rev = version;
```

SuperSandro2000 commented on pkgs/servers/monitoring/prometheus/sachet.nix:
```diff
@@ -0,0 +1,26 @@
+      license = licenses.bsd2;
+      maintainers = with maintainers; [ govanify ];
+    };
+  };
+}
```
```suggestion
  };
```

SuperSandro2000 reviewed (changes requested):
Please reformat the file.

GovanifY commented:
@SuperSandro2000 reformatted using nixpkgs-fmt and implemented your suggestions

SuperSandro2000 commented:
I think you also want to add service integration for this.

GovanifY commented:
Yes, I was waiting first to get this PR accepted/reviewed before proceeding with the service. I'm going to work on creating the service in this PR then

GovanifY committed:
nixos/prometheus.sachet: add module

GovanifY commented:
I've added the module, here is the configuration to test it:

```nix
{ config, pkgs, lib, ... }:
{
  networking = {
    hostName = "sachet";
  };
  users = {
    mutableUsers = false;
    extraUsers = {
      root = {
        password = "toor";
      };
    };
  };
  services.prometheus.sachet = {
    enable = true;
    configuration = {
      providers = {
        twilio = {
          account_sid = "aCb3bbaacc554b";
          auth_token = "4736473aaabbcc66";
        };
      };
      templates = [ ./some-template.tmpl ];
      receivers = [{
        name = "pager";
        provider = "twilio";
        to = [ "+33123456789" ];
        text = "{{ template \"telegram_message\" . }}";
      }];
    };
  };
}
```

NB: you might want to either remove text and templates or copy paste a template to test it in the configuration, I've just tested the default configuration I've added to the module for consistency sake.

And here's the log shown on my testing VM:

![screenshot](https://user-images.githubusercontent.com/6375438/125159624-1334ae00-e168-11eb-935c-374a37c1b66d.png)

SuperSandro2000 reviewed (commented):
Can you re-use the default export options?

https://github.com/NixOS/nixpkgs/blob/53ca0b16646ae358e37c396f6c5f8955977002c1/nixos/modules/services/monitoring/prometheus/exporters.nix#L51

https://github.com/NixOS/nixpkgs/blob/53ca0b16646ae358e37c396f6c5f8955977002c1/nixos/modules/services/monitoring/prometheus/exporters/pihole.nix

GovanifY commented:
@SuperSandro2000 This isn't a prometheus exporter though, it's meant to be used with webhooks and doesn't scrape anything, as exporter does, it only exposes SMS providers to prometheus. I put it under prometheus as it's rigged up to reply to alertmanager webhooks queries only.

SuperSandro2000 commented:
I didn't realize that.

SuperSandro2000 commented on nixos/modules/services/monitoring/prometheus/sachet.nix:
```diff
@@ -0,0 +1,79 @@
+        example = literalExample ''
+          {
+            providers = {
+              twilio = {
+                account_sid = "aCb3bbaacc554b";
+                auth_token = "4736473aaabbcc66";
+              };
```
I don't feel great about putting credentials into the nix store.

GovanifY replied:
Hey, I'd like to revisit this PR. Unfortunately the configuration file has no way to specify the use of any environment variable. As such I don't see how I can avoid doing this.

GovanifY replied:
Actually it does parse environment variables by default, so I've added that instead in the example!

GovanifY replied:
I've homebrewed a way with envsubst that seems to work.

GovanifY commented:
Hey, I've revisited this PR, bumped the version and took care of the last review comments. I would appreciate another look at it!

winterqt commented on pkgs/servers/monitoring/prometheus/sachet.nix:
```diff
@@ -0,0 +1,25 @@
+  vendorSha256 = null;
+
+  meta = with lib; {
+    description = "Sachet is an SMS alerting tool for the Prometheus Alertmanager.";
```
Don't include a period, and don't repeat the name of the package -- see [the manual](https://nixos.org/manual/nixpkgs/stable/#sec-standard-meta-attributes).

SuperSandro2000 commented on nixos/modules/services/monitoring/prometheus/sachet.nix:
```diff
@@ -0,0 +1,86 @@
+  cfg = config.services.prometheus.sachet;
+  configFile = pkgs.writeText "sachet.yml" (builtins.toJSON cfg.configuration);
+
+  # sachet does not allow parsing of environment variables in its configuration
+  # file, so let's homebrew our own!
```
```suggestion
  # sachet does not allow parsing of environment variables in its configuration file
```

SuperSandro2000 commented on nixos/modules/services/monitoring/prometheus/sachet.nix:
```diff
@@ -0,0 +1,86 @@
+        '';
+      };
+
+      listenAddress = mkOption {
+        type = types.str;
+        default = "localhost:9876";
+        description = ''
+          The address sachet will listen to.
```
This is address and port. Usually we separate port.

SuperSandro2000 commented on nixos/modules/services/monitoring/prometheus/sachet.nix:
```diff
@@ -0,0 +1,86 @@
+    }) (mkIf cfg.enable {
+        systemd.services.sachet = {
+          wantedBy = [ "multi-user.target" ];
+          after = [ "network-online.target" ];
```
```suggestion
          after = [ "network.target" "network-online.target" ];
```

GovanifY commented:
Thanks for the review! took care of the comments

GovanifY commented:
Hi @SuperSandro2000 , I took care of your comments, would you mind taking another look at this PR?

winterqt commented on nixos/modules/services/monitoring/prometheus/sachet.nix:
```diff
@@ -0,0 +1,92 @@
+  configFile = pkgs.writeText "sachet.yml" (builtins.toJSON cfg.configuration);
+
+  # sachet does not allow parsing of env variables in its configuration file
+  runScript = pkgs.writeShellScript "run.sh" ''
```
```suggestion
  runScript = pkgs.writeShellScript "sachet-run" ''
```

winterqt commented on nixos/modules/services/monitoring/prometheus/sachet.nix:
```diff
@@ -0,0 +1,92 @@
+  # sachet does not allow parsing of env variables in its configuration file
+  runScript = pkgs.writeShellScript "run.sh" ''
+    ${pkgs.envsubst}/bin/envsubst -i "${configFile}" > /tmp/sachet.yaml
+    ${pkgs.prometheus-sachet}/bin/sachet -config /tmp/sachet.yaml -listen-address $1
```
Prefer `exec` here for semantic reasons, and double-quoting `$1` because ShellCheck says to ([SC2086](https://www.shellcheck.net/wiki/SC2086)).

```suggestion
    exec ${pkgs.prometheus-sachet}/bin/sachet -config /tmp/sachet.yaml -listen-address "$1"
```

winterqt commented on nixos/modules/services/monitoring/prometheus/sachet.nix:
```diff
@@ -0,0 +1,92 @@
+          }
+        '';
+        description = ''
+          Sachet configuration's as a nix attribute set.
```
Stole this from Postgres, maybe it sounds better?

```suggestion
          Sachet's configuration, specified as an attribute set.
```

winterqt commented on nixos/modules/services/monitoring/prometheus/sachet.nix:
```diff
@@ -0,0 +1,92 @@
+        '';
+      };
+
+      address = mkOption {
+        type = types.str;
+        default = "localhost";
+        description = ''
+          The address Sachet will listen to.
+        '';
+      };
```
Shouldn't this be `host`, as `address` would imply a host + a port (when they're separate here)?

GovanifY replied:
address and port are actually used together and sandro told me to prefer this naming scheme.

winterqt commented on nixos/modules/services/monitoring/prometheus/sachet.nix:
```diff
@@ -0,0 +1,92 @@
+  config = mkIf cfg.enable {
+    assertions = singleton {
+      assertion = cfg.configuration != null;
+      message = "Can not enable sachet without a configuration.";
```
```suggestion
      message = "Cannot enable Sachet without a configuration.";
```

GovanifY commented:
Hey @SuperSandro2000 , this PR should be good to merge by this point, could you please take another look?

winterqt commented:
(Pushed to remove extraneous `Type = "simple"`, didn't want to bother you any more.)

Waiting on OfBorg, sorry this took so long.

pennae commented:
fyi, this seems to have introduced a new warning in the docs build:
```
trace: warning: literalExample is deprecated, use literalExpression instead, or use literalDocBook for a non-Nix description.
```
